document/JS.bs

```diff
@@ -0,0 +1,921 @@
+    1. If |module| is [=error=], return false.
+    1. Return true.
+</div>
+
+<div algorithm>
+  To <dfn>construct a WebAssembly module object</dfn> from a module |module|, perform the following steps:
+
+    1. Let |moduleObject| be a new {{Module}} object.
+    1. Set |moduleObject|.\[[Module]] to |module|.
+    1. Set |moduleObject|.\[[Bytes]] to |bytes|.
+    1. Return |moduleObject|.
+</div>
+
+<div algorithm>
+  To <dfn>asynchronously compile a WebAssembly module</dfn> from a {{BufferSource}} |bytes|, with the promise |promise|, using optional [=task source=] |taskSource|, perform the following steps:
+
+    1. In parallel, [=compile a WebAssembly module|compile the WebAssembly module=] |bytes| and store the result as |module|.
+    1. When the above operation completes, [=queue a task=] to perform the following steps. If |taskSource| was provided, queue the task on that task source.
+        1. If |module| is [=error=], reject |promise| with a {{CompileError}} exception.
+        1. Otherwise,
+            1. [=Construct a WebAssembly module object=] from |module| and let |moduleObject| be the result.
+            1. [=Resolve=] |promise| with |moduleObject|.
+</div>
+
+<div algorithm>
+    The <dfn method for="WebAssembly">compile(bytes)</dfn> method, when invoked, performs the following steps:
+    1. Let |stableBytes| be a [=get a copy of the buffer source|copy of the bytes held by the buffer=] |bytes|.
+    1. Let |promise| be [=a new promise=].
+    1. [=Asynchronously compile a WebAssembly module=] from |stableBytes| with |promise|.
+    1. Return |promise|.
+</div>
+
+<div algorithm>
+  To <dfn>instantiate a WebAssembly module</dfn> from a {{Module}} |moduleObject| and imports |importObject|, perform the following steps:
+    1. Let |module| be |moduleObject|.\[[Module]].
+    1. If |module|.imports is not an empty list, and |importObject| is undefined, throw a [=TypeError=] exception.
+    1. Let |funcs| be an empty [=list=] of callable JavaScript objects.
+    1. Let |memories| be an empty [=list=] of {{Memory}} objects.
+    1. Let |tables| be an empty [=list=] of {{Table}} objects.
+
+    Note: The |funcs|, |memories| and |tables| lists are collected and used in order to cache the JavaScript objects corresponding to WebAssembly objects. If a WebAssembly object is exported multiple times, it will appear in exports as the same object instance. Moreover, if a JavaScript wrapper of a WebAssembly object was imported, and then present on the export, a new wrapper will not need to be created. However, if an ordinary JS function is imported, it will appear as wrapped by an {{ExportedFunction}}.
+
+    6. Let |imports| be an empty [=list=] of [=external value=]s.
+    1. For each (|moduleName|, |componentName|, |externtype|) in [=module_imports=](|module|), do
+        1. Let |o| be ? [=Get=](|importObject|, |moduleName|).
```

Comment:
Again, Web IDL has covered this for you. importObject is not a JS object anymore, but instead an ordered map of (JS string ->(JS string -> (double or ExportedFunction or Memory or Table)).

Reply:
Hmm, this will definitely be an observable semantic change. In particular, we'd no longer search up the prototype chain. I'm not sure how bad that would be, given the point Wasm is at in shipping.

Reply:
Why wouldn't you search up the prototype chain? Web IDL definitely does that, during the initial conversion steps.

Reply:
The [conversion algorithm for records](https://heycam.github.io/webidl/#es-record) seems to be based on reading all the own property keys as the first step. It seems like this would reach a different result than the algorithm here, which looks at the module to find what to look for and does a simple Get on those, regardless of whether they are own or not. Currently, I have InstanceImportsMap as a record, though maybe there's a better type.

Reply:
Oh, sorry about that, I guess I just totally forgot how records worked.

If you do need to drop down to JS, you can use the `object` type. I'm not sure what would be cleaner; it seems kind of nice to do all the JS object stuff ahead of time in one fell swoop then operate on the unobservable spec/implementation data structures, but on the other hand it sounds like that might involve unnecessary work, and changing implementations is annoying.

The ahead-of-time version seems a lot nicer for the async algorithm version (not yet in this document).

I haven't checked that the previous spec was well-followed or detailed enough before your version; that might also be worth investigating. E.g. maybe someone is already using a record-like algorithm today! (We could test with proxies.)

Anyway, the above is just a lot of thoughts. I don't have strong feelings, just hoping we weigh all the options.

Reply:
I asked at the WebAssembly WG meeting about asynchronous instantiation, and the response was pretty unequivocal: no, that would not make sense. I'll switch to `object` here for now; I filed a WebIDL bug about [another feature mismatch](https://github.com/heycam/webidl/issues/469).

Reply:
I meant for instantiateStreaming. The question is do we delay grabbing object properties until we post a task back to the main thread, or do we do it immediately when the function is called, and let the later work operate without worrying about JS objects. In the former case, we could e.g. change the object's properties while the Response was coming from the network, and expect that to be detected; for the latter, doing so would have no effect.

Reply:
@littledan the IDL bug you point to is about return values, but this seems to be about something else?